tsdb/index/index.go

```diff
@@ -1164,53 +1164,50 @@ func newReader(b ByteSlice, c io.Closer) (*Reader, error) {
 		// Earlier V1 formats don't have a sorted postings offset table, so
 		// load the whole offset table into memory.
 		r.postingsV1 = map[string]map[string]uint64{}
-		if err := ReadOffsetTable(r.b, r.toc.PostingsTable, func(key []string, off uint64, _ int) error {
-			if len(key) != 2 {
-				return errors.Errorf("unexpected key length for posting table %d", len(key))
+		if err := ReadOffsetTable(r.b, r.toc.PostingsTable, PostingTableReader, func(key labels.Label, off uint64, _ int) error {
+			if _, ok := r.postingsV1[key.Name]; !ok {
+				r.postingsV1[key.Name] = map[string]uint64{}
+				r.postings[key.Name] = nil // Used to get a list of labelnames in places.
 			}
-			if _, ok := r.postingsV1[key[0]]; !ok {
-				r.postingsV1[key[0]] = map[string]uint64{}
-				r.postings[key[0]] = nil // Used to get a list of labelnames in places.
-			}
-			r.postingsV1[key[0]][key[1]] = off
+			r.postingsV1[key.Name][key.Value] = off
 			return nil
 		}); err != nil {
 			return nil, errors.Wrap(err, "read postings table")
 		}
 	} else {
-		var lastKey []string
+		var lastKey labels.Label
+		var lastSet bool
 		lastOff := 0
 		valueCount := 0
 		// For the postings offset table we keep every label name but only every nth
 		// label value (plus the first and last one), to save memory.
-		if err := ReadOffsetTable(r.b, r.toc.PostingsTable, func(key []string, _ uint64, off int) error {
-			if len(key) != 2 {
-				return errors.Errorf("unexpected key length for posting table %d", len(key))
-			}
-			if _, ok := r.postings[key[0]]; !ok {
+		if err := ReadOffsetTable(r.b, r.toc.PostingsTable, PostingTableReader, func(key labels.Label, _ uint64, off int) error {
+			if _, ok := r.postings[key.Name]; !ok {
 				// Next label name.
-				r.postings[key[0]] = []postingOffset{}
-				if lastKey != nil {
+				r.postings[key.Name] = []postingOffset{}
+				if lastSet {
 					// Always include last value for each label name.
-					r.postings[lastKey[0]] = append(r.postings[lastKey[0]], postingOffset{value: lastKey[1], off: lastOff})
+					r.postings[lastKey.Name] = append(r.postings[lastKey.Name], postingOffset{value: lastKey.Value, off: lastOff})
 				}
-				lastKey = nil
+				lastKey = labels.Label{}
```

Comment:
Even if not strictly required, I would also set `lastSet = false` to avoid future regressions if the code below will be changed.

Reply:
[I removed this assignment instead](https://github.com/prometheus/prometheus/pull/11535/commits/e83ee0218cfd092056c437aaad67189fc396cc58), because both lastKey and lastSet are updated in the next few lines.